### PR TITLE
[eas-cli] Enable autoIncrementing on Android

### PR DESCRIPTION
**Pull request by jkhales**
# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary.
- [ ] I've tagged the changelog entry with `[EAS BUILD API]` if it's a part of a breaking change to EAS Build API (only possible when updating `@expo/eas-build-job` package).

# Why

extend auto incrementing to Android.

# How

Followed this PR: https://github.com/expo/eas-cli/pull/231

# Test Plan

Added tests.

Ran a a managed and bare demo repo with `autoIncrement : true` and confirmed that the appropriate files were mutated.

### CODE REVIEW

File: packages/eas-cli/src/build/android/version.ts [hunk 1]
```diff
@@ -2,13 +2,92 @@ import { ExpoConfig } from '@expo/config';
...
+      return;
+    }
+    case BumpStrategy.SHORT_VERSION: {
+      const shortVersion = AndroidConfig.Version.getVersionName(exp) ?? '1.0.0';
+      await bumpShortVersionAsync({ shortVersion, projectDir, exp });
+      break;
+    }
+    case BumpStrategy.VERSION_CODE:
+    default: {
+      const versionCode = AndroidConfig.Version.getVersionCode(exp);
```
Comment (jkhales) [inline packages/eas-cli/src/build/android/version.ts:77]: `getVersionCode` returns a `number` so we don't need to do the semver parsing done for the iOS `buildNumber`.

File: packages/eas-cli/src/build/utils/version.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import { updateAppJsonConfigAsync } from './appJson';
+
+export function ensureStaticConfigExists(projectDir: string): void {
+  const paths = getConfigFilePaths(projectDir);
+  if (!paths.staticConfigPath) {
+    throw new Error('autoIncrement option is not supported when using app.config.js');
+  }
+}
+
+export async function bumpShortVersionAsync({
```
[the remark this answers is not in the document]
Comment (dsokal) [reply packages/eas-cli/src/build/utils/version.ts:17]: Maybe `bumpShortVersion` -> `bumpVersionAsync` and `shortVersion` -> `version`? Could you update the name in other places as well (e.g. `BumpStrategy` enums)?

File: packages/eas-cli/src/build/android/version.ts [hunk 3]
```diff
@@ -39,3 +118,39 @@ export async function maybeResolveVersionsAsync(
...
+  writeBuildGradle({ projectDir, buildGradle: updatedBuildGradle });
+  return updatedBuildGradle;
+}
+
+export function readBuildGradle(projectDir: string): string | undefined {
+  const buildGradlePath = AndroidConfig.Paths.getAppBuildGradleFilePath(projectDir);
+  if (!fs.pathExistsSync(buildGradlePath)) {
+    return undefined;
+  }
+  return fs.readFileSync(buildGradlePath, 'utf8');
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:144]:
```suggestion
  return await fs.readFile(buildGradlePath, 'utf8');
```
Don't use sync functions in node if there are async counterparts and you really don't have to use sync functions.

File: packages/eas-cli/src/build/android/version.ts [hunk 4]
```diff
@@ -39,3 +118,39 @@ export async function maybeResolveVersionsAsync(
...
+
+function writeBuildGradle({
+  projectDir,
+  buildGradle,
+}: {
+  projectDir: string;
+  buildGradle: string;
+}): void {
+  const buildGradlePath = AndroidConfig.Paths.getAppBuildGradleFilePath(projectDir);
+  fs.writeFileSync(buildGradlePath, buildGradle);
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:155]: Same here.

File: packages/eas-cli/src/build/android/version.ts [hunk 5]
```diff
@@ -39,3 +118,39 @@ export async function maybeResolveVersionsAsync(
...
+async function writeVersionsToBuildGradleAsync({
+  projectDir,
+  exp,
+}: {
+  projectDir: string;
+  exp: ExpoConfig;
+}): Promise<string> {
+  const buildGradle = readBuildGradle(projectDir);
+  if (!buildGradle) {
+    throw new Error('missing build gradle');
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:131]: The error message is user-facing so please make it more friendly. Use full sentences. 
Also `build gradle` -> `build.gradle`

File: packages/eas-cli/src/build/android/version.ts [hunk 6]
```diff
@@ -2,13 +2,92 @@ import { ExpoConfig } from '@expo/config';
...
+    case BumpStrategy.NOOP: {
+      return;
+    }
+    case BumpStrategy.SHORT_VERSION: {
+      const shortVersion = AndroidConfig.Version.getVersionName(exp) ?? '1.0.0';
+      await bumpShortVersionAsync({ shortVersion, projectDir, exp });
+      break;
+    }
+    case BumpStrategy.VERSION_CODE:
+    default: {
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:76]: I don't see a benefit from using `switch` over `if`. With `if`, you wouldn't need to handle `default` and I think the code would be easier to read.

Comment (dsokal) [reply packages/eas-cli/src/build/android/version.ts:76]:
```
if (bumpStrategy === BumpStrategy.VERSION) {
  ...
} else if (bumpStrategy === BumpStrategy.VERSION_CODE) { 
  ...
} 
```

File: packages/eas-cli/src/build/android/version.ts [hunk 7]
```diff
@@ -2,13 +2,92 @@ import { ExpoConfig } from '@expo/config';
...
 } from '../../project/android/gradleUtils';
 import { resolveWorkflowAsync } from '../../project/workflow';
+import { updateAppJsonConfigAsync } from '../utils/appJson';
+import { bumpShortVersionAsync, ensureStaticConfigExists } from '../utils/version';
+
+export enum BumpStrategy {
+  SHORT_VERSION,
+  VERSION_CODE,
+  NOOP,
+}
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:22]:
```suggestion
}

```

**Review by dsokal — CHANGES_REQUESTED** [review]

File: packages/eas-cli/src/build/android/version.ts [hunk 8]
```diff
@@ -2,13 +2,92 @@ import { ExpoConfig } from '@expo/config';
...
+}): Promise<void> {
+  if (bumpStrategy === BumpStrategy.NOOP) {
+    return;
+  }
+  ensureStaticConfigExists(projectDir);
+
+  const buildGradle = await getAppBuildGradleAsync(projectDir);
+  const isMultiFlavor =
+    buildGradle.android?.productFlavors ||
+    (buildGradle.android?.flavorDimensions ?? '').split(',').length > 1;
```
Comment (wkozyra95) [inline packages/eas-cli/src/build/android/version.ts:40]:
```suggestion
    buildGradle.android?.flavorDimensions;
```
nit pick: without productFlavors flavorDimensions does not make sense

**Review by wkozyra95 — COMMENTED** [review]
remember to update docs https://docs.expo.dev/build/eas-json/#android-specific-options

File: packages/eas-cli/src/build/android/version.ts (hunk 6 above)
Comment (jkhales) [reply packages/eas-cli/src/build/android/version.ts:76]: > I don't see a benefit from using switch over if

I'm happy to change it, but isn't this a canonical use case for `switch`, are you just opposed to `switch` in general?

Comment (dsokal) [reply packages/eas-cli/src/build/android/version.ts:76]: - (I don't like `switch` in general, but it's just my opinion)
- Even if I liked using `switch`, I think it does not makes sense in this case. There are only two states that you need to handle

File: packages/eas-cli/src/build/utils/version.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkhales) [reply packages/eas-cli/src/build/utils/version.ts:17]: `bumpVersionAsync` is taken, thinking to do `bumpAppVersionAsync` instead following the description here: https://docs.expo.dev/versions/latest/config/app/#version

**Comment by jkhales** [comment]
> remember to update docs https://docs.expo.dev/build/eas-json/#android-specific-options

thanks for the reminder, updated! https://github.com/expo/expo/commit/0c42338ad2bea4368be8380058dd7ea3f89ec6b7

File: packages/eas-cli/src/build/utils/version.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+export function ensureStaticConfigExists(projectDir: string): void {
+  const paths = getConfigFilePaths(projectDir);
+  if (!paths.staticConfigPath) {
+    throw new Error('autoIncrement option is not supported when using app.config.js');
+  }
+}
+
+export async function bumpAppVersionAsync({
+  appVersion: shortVersion,
```
Comment (dsokal) [inline packages/eas-cli/src/build/utils/version.ts:18]:
```suggestion
  appVersion,
```

File: packages/eas-cli/src/build/utils/version.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  projectDir,
+  exp,
+}: {
+  appVersion: string;
+  projectDir: string;
+  exp: ExpoConfig;
+}): Promise<void> {
+  let bumpedAppVersion: string;
+  if (semver.valid(shortVersion)) {
```
Comment (dsokal) [inline packages/eas-cli/src/build/utils/version.ts:27]:
```suggestion
  if (semver.valid(appVersion)) {
```

File: packages/eas-cli/src/build/utils/version.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    bumpedAppVersion = nullthrows(semver.inc(shortVersion, 'patch'));
```
Comment (dsokal) [inline packages/eas-cli/src/build/utils/version.ts:28]:
```suggestion
    bumpedAppVersion = nullthrows(semver.inc(appVersion, 'patch'));
```

File: packages/eas-cli/src/build/utils/version.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    Log.log(
+      `Bumping ${chalk.bold('expo.version')} from ${chalk.bold(shortVersion)} to ${chalk.bold(
```
Comment (dsokal) [inline packages/eas-cli/src/build/utils/version.ts:30]:
```suggestion
      `Bumping ${chalk.bold('expo.version')} from ${chalk.bold(appVersion)} to ${chalk.bold(
```

File: packages/eas-cli/src/build/utils/version.ts [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        bumpedAppVersion
+      )}`
+    );
+  } else {
+    Log.log(`${chalk.bold('expo.version')} = ${chalk.bold(shortVersion)} is not a valid semver`);
```
Comment (dsokal) [inline packages/eas-cli/src/build/utils/version.ts:35]:
```suggestion
    Log.log(`${chalk.bold('expo.version')} = ${chalk.bold(appVersion)} is not a valid semver`);
```

File: packages/eas-cli/src/build/android/version.ts [hunk 14]
```diff
@@ -2,13 +2,89 @@ import { ExpoConfig } from '@expo/config';
...
+  if (isMultiFlavor) {
+    throw new Error('automatic version bumping is not supported for multiflavor android projects');
```
[the remark this answers is not in the document]
Comment (dsokal) [reply packages/eas-cli/src/build/android/version.ts:42]:
```suggestion
    throw new Error('Automatic version bumping is not supported for multi-flavor Android projects.');
```

File: packages/eas-cli/src/build/android/version.ts [hunk 15]
```diff
@@ -2,13 +2,84 @@ import { ExpoConfig } from '@expo/config';
...
+}: {
+  bumpStrategy: BumpStrategy;
+  projectDir: string;
+  exp: ExpoConfig;
+}): Promise<void> {
+  ensureStaticConfigExists(projectDir);
+  Log.addNewLineIfNone();
+
+  if (bumpStrategy === BumpStrategy.NOOP) {
+  } else if (bumpStrategy === BumpStrategy.APP_VERSION) {
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:67]:
```suggestion
  if (bumpStrategy === BumpStrategy.APP_VERSION) {
```

File: packages/eas-cli/src/build/android/version.ts [hunk 16]
```diff
@@ -2,13 +2,84 @@ import { ExpoConfig } from '@expo/config';
...
+
+export async function bumpVersionInAppJsonAsync({
+  bumpStrategy,
+  projectDir,
+  exp,
+}: {
+  bumpStrategy: BumpStrategy;
+  projectDir: string;
+  exp: ExpoConfig;
+}): Promise<void> {
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:62]:
```suggestion
}): Promise<void> {
  if (bumpStrategy === BumpStrategy.NOOP) {
    return;
  }
```

File: packages/eas-cli/src/build/android/version.ts [hunk 17]
```diff
@@ -39,3 +110,39 @@ export async function maybeResolveVersionsAsync(
...
+  projectDir: string;
+  exp: ExpoConfig;
+}): Promise<string> {
+  const buildGradle = await readBuildGradleAsync(projectDir);
+  if (!buildGradle) {
+    throw new Error('This project is missing a build.gradle file.');
+  }
+  let updatedBuildGradle = AndroidConfig.Version.setVersionName(exp, buildGradle);
+  updatedBuildGradle = AndroidConfig.Version.setVersionCode(exp, updatedBuildGradle);
+  writeBuildGradleAsync({ projectDir, buildGradle: updatedBuildGradle });
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:127]:
```suggestion
  await writeBuildGradleAsync({ projectDir, buildGradle: updatedBuildGradle });
```

File: packages/eas-cli/src/build/android/version.ts [hunk 18]
```diff
@@ -2,13 +2,84 @@ import { ExpoConfig } from '@expo/config';
...
+  projectDir: string;
+  exp: ExpoConfig;
+}): Promise<void> {
+  ensureStaticConfigExists(projectDir);
+  Log.addNewLineIfNone();
+
+  if (bumpStrategy === BumpStrategy.NOOP) {
+  } else if (bumpStrategy === BumpStrategy.APP_VERSION) {
+    const shortVersion = AndroidConfig.Version.getVersionName(exp) ?? '1.0.0';
+    await bumpAppVersionAsync({ appVersion: shortVersion, projectDir, exp });
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:69]:
```suggestion
    const appVersion = AndroidConfig.Version.getVersionName(exp) ?? '1.0.0';
    await bumpAppVersionAsync({ appVersion, projectDir, exp });
```

File: packages/eas-cli/src/build/ios/version.ts [hunk 19]
```diff
@@ -52,29 +51,9 @@ export async function bumpVersionInAppJsonAsync({
...
-      const { bumpedShortVersion } = await promptAsync({
-        type: 'text',
-        name: 'bumpedShortVersion',
-        message: 'What is the next version?',
-      });
-      await updateAppJsonConfigAsync({ projectDir, exp }, config => {
-        config.version = bumpedShortVersion;
-      });
-    }
+    await bumpAppVersionAsync({ appVersion: shortVersion, projectDir, exp });
```
Comment (dsokal) [inline packages/eas-cli/src/build/ios/version.ts:56]:
```suggestion
    const appVersion = IOSConfig.Version.getVersion(exp);
    await bumpAppVersionAsync({ appVersion, projectDir, exp });
```

**Review by dsokal — APPROVED** [review]
Just a few nits.

**Comment by dsokal** [comment]
@jkhales 

> > remember to update docs https://docs.expo.dev/build/eas-json/#android-specific-options
>
> thanks for the reminder, updated! expo/expo@0c42338

It should have been committed after we merge and publish eas-cli to npm. Right now, if you read the docs (https://docs.expo.dev/build/eas-json/#autoincrement) you couldn't use it in your project. Users can be surprised it's not working.

**Comment by jkhales** [comment]
> @jkhales
> 
> > > remember to update docs https://docs.expo.dev/build/eas-json/#android-specific-options
> > 
> > 
> > thanks for the reminder, updated! [expo/expo@0c42338](https://github.com/expo/expo/commit/0c42338)
> 
> It should have been committed after we merge and publish eas-cli to npm. Right now, if you read the docs (https://docs.expo.dev/build/eas-json/#autoincrement) you couldn't use it in your project. Users can be surprised it's not working.

Ah woops. 😖

**Pull request merged**